Scientific Question: For animals that are known to be less susceptible to cancer, would they have similar p53 protein sequences as human p53 because human p53 is an important tumor suppressor protein that reduces the risk of cancer in humans?

Background: In humans, p53 is an important tumor suppressor protein that reduces the risk for cancer. P53 is discovered in 1979 and is made by a gene known as TP53. P53 prevents cells from growing and dividing too fast or in an uncontrolled way. P53 is located in the nucleus of cells throughout the body. The loss of p53 can lead to the multiplication of cancer cells even when the cancer cells do not have growth stimuli (Benedict 2018). Therefore, having p53 is crucial in preventing cancer and it plays a very significant role in lowering the risk for tumors and cancer. The p53 sequences of human, chimpanzee, dogs, and elephants used in this code are obtained from NCBI. All the p53 sequences are downloaded from NCBI as fasta files. 

Scientific Hypothesis: If an animal is known to be less susceptible to cancer, then it will contain similar p53 sequence as humans because p53 is an important tumor suppressor protein in humans.

In order to test the scientific hypothesis, two analyses were done using bioinfomatics method of multiple sequence alignment and pairwise sequence alignment. Then a histogram is created to visualize the perent identity between human-chimpanzee p53, human-dogs p53, and human-elephants p53. Multiple sequence alignment (msa) is performed to generate an alignment of multiple sequences, four in this case, so we can see where the sequences are similar. MSA can help us assess sequence conservation of protein domains, infer homology, and infer evolutionary relationships among the aligned sequences. On the other hand, pairwise sequence alignment can align two or more sequences as well, but pairwise sequence alignment uses a scoring system where the alignment score would be matches (+1) or mismatches (-1) between sequences. Pairwise sequence alignment is used to identify the regions of similarity that may indicate significant functional, structural, and evolutionary relationships between the aligned sequences. In this case, msa of the four sequences (human p53, chimpanzee p53, dogs p53, elephants p53) are done. Three Pairwise sequence alignments (human and chimpanzee p53, human and dogs p53, human and elephants p53) are done. The necessary data (p53 sequences of human, chimpanzee, dogs, elephants) are downloaded from NCBI.

I used 8 packages in total for performing the entire code. Before loading and using the packages, I need to install the packages first by using install.packages(). Here, I will define all the packages used in the code. 

1) seqinr: this package is responsible to perform data analysis and data management. Most used function from this package would be read.fasta where it reads the nucleotide or amino-acid sequences from a file in FASTA format.

2) BiocManager: this package provide functions to perform statistical analysis and prepare genomic data. This package is easy for everyday use and its install function helps helps me accurately install the necessary packages.

3) Biostrings: this package is used for containing and manipulating strings and string algorithms. With this package, I can easily manipulate large sequences or sets of sequences. Useful functions like readAAStringSet() helps me read the protein sequences from the fasta file.

4) msa: the package that provides an interface to three multiple sequence alignment algorithms (ClustalW, ClustalOmega, MUSCLE). By using the msa function in the package, a multiple sequence alignment of multiple sequences can be performed.

5) ggmsa: the package that is designed for the visualization of multiple sequence alignment. This package can visualize and annotate the output of multiple sequence alignment. 

6) tinytex: a package for managing LaTeX data, such as compiling a LaTeX document to a PDF. With this package, we can use the msaPrettyPrint function to visualzie the output of multiple sequence alignment.  

7) bio3d: a package that is designed for the analysis of protein structure and sequence data. The seqidentity function in the package can determine the percent identity between two aligned sequences, which is extremely helpful. 

8) DECIPHER: a package that is designed for deciphering and managing sequences. The BrowseSeqs function in the package shows the aligned sequences in XStringSet. BrowseSeqs function is very helpful in visualizing the string set after pairwise sequence alignment.


```{r}

# this step loads all the necessary packages
library(seqinr)
library(BiocManager)
library(Biostrings)
library(msa)
library(ggmsa)
library(tinytex)
library(bio3d)
library(DECIPHER)

# Here, I perform multiple sequence alignment (msa) of four sequences (sequences of human p53, chimpanzee p53, dogs p53, elephants p53). Msa is a method where it generates an alignment of multiple sequences, so I can see where the sequences are similar. Msa helps me assess sequence conservation of protein domains, infer homology, and infer evolutionary relationships among the aligned sequences.

# In order to have the method work, the first thing we need to do is have a fasta file that contains all the sequences we want to align. Here, "all.sequence.fasta.txt" is the fasta file with the four sequences that I want to align. I read the amino acid string set in the file and stored it in the variable called all_sequences_fasta.
all_sequences_fasta <- readAAStringSet("all.sequence.fasta.txt")
all_sequences_fasta

# Now that I have read the amino acid string set of all four sequences, I would perform multiple sequence alignment of the sequences using the msa function in the msa package. The msa function takes in input sequences, which can be a character vector, an object of class XStringSet, or a single character string with a file name. Here, the msa function takes in the protein string set we read in the last step and perform multiple sequence alignment of the sequences.
all_sequences_align <- msa(all_sequences_fasta)
all_sequences_align

# Here I print the four completely aligned sequences in the console.
print(all_sequences_align, show="complete")

# This is one method to visualize the output of multiple sequence alignment, which is by using the msaPrettyPrint function. The msaPrettyPrint is a function in the msa package where we first pass it the msa output from all_sequences_align. The output type here is tex, stands for LaTeX, but there are other output options as well. In order to have the code to work, I have to first install the tinytex package that manages LaTeX data. The LaTeX file made by this code can be found in the Files tab. Then when I click compile PDF, tinytex would compile the file and create a PDF visualizing the output of msa. 
msaPrettyPrint(all_sequences_align, output="tex", showNames="left",
               showLogo="top", askForOverwrite=FALSE, verbose=FALSE)

# This is another method for visualizing msa. The function ggmsa is found in the ggmsa package, which is designed for msa visualization. The function ggmsa takes in all_sequences_fasta, which is the protein string set of all the sequences. There are additional parameters such as the start amino acid and the end amino acid, font, and color. This function is very easy to use that creates appealing msa outputs. 
all_sequences_plot <- ggmsa(all_sequences_fasta, 300, 350, color = "Clustal", font = "DroidSansMono", char_width = 0.5, seq_name = TRUE )



# After performing msa, I would perform pairwise sequence alignment now. Pairwise sequence alignment can align two or more sequences and uses a scoring system determining the alignment score matches (+1) and mismatches (-1) between sequences. Pairwise sequence alignment is used to identify the regions of similarity that may indicate significant functional, structural, and evolutionary relationships between the aligned sequences. 

#I am performing three separate pairwise sequence alignments here instead of a single pairwise sequence for all four sequences because my scientific hypothesis is that if an animal is known to be less susceptible to cancer, then it will contain similar p53 sequence as humans. Therefore, I want to know the alignment result and percent identity between human and chimpanzee p53, human and dogs p53, and humans and elephants p53. 

# Here, I are performing a pairwise sequence alignment of p53 for humans and chimpanzees. The method works in a couple steps. First, I need to have a fasta file containing the two sequences I want to align, which is human and chimpanzee p53 in this case. The function readAAStringSet takes in this string set argument and the result is stored in the variable human_chimps_fasta. The variable human_chimps_fasta is an AAStringSet variable or an S4 variable that is a global variable. Then the function read.fasta will simily read the fasta file "human_chimp.fasta.txt" and creates a list variable that is a global variable. Global variables are the variables that are defined outside of the functions and can be accessed everywhere in the program. Local variables are the variables that are defined in a method and cannot be accessed outside of that method. 

# human_chimps_read is a list varibale, so the length function can be used to determine the numer of sequences. I put length(human_chimps_read)-1 here because there was a small error in my read.fasta function where I have two sequences in my fasta file yet the read.fasta function reads the file and says that there are three sequences. Therefore, I put -1 here in order to keep my code away from out of bound. 
human_chimps_fasta <- readAAStringSet("human_chimp.fasta.txt")
human_chimps_read <- read.fasta(file = "human_chimp.fasta.txt")
human_chimps_length <- 1:(length(human_chimps_read)-1)

# Here, I create a function to perform pairwise sequence alignment for two sequences. If there are more than two sequences, a nested for loop should be used. The function name is seq_pairwise and takes in two arguments of sequences and seq_length. 
seq_pairwise <- function(sequences, seq_length) {
  
      #define the two string variables for two sequences 
      string1 <- sequences[1]
      string2 <- sequences[2]

      #Turning the string1 variable into a character string
      #First unlisting the protein string set list then turning it into                 character string
      list_stringset_obj = AAStringSetList(string1)
      as.character(unlist(list_stringset_obj))
      as_character_list = as(list_stringset_obj, "CharacterList")
      as.list(as_character_list)
      
      #Turning the string2 variable into a character string
       #First unlisting the protein string set list then turning it into                 character string
      list_stringset_obj2 = AAStringSetList(string2)
      as.character(unlist(list_stringset_obj2))
      as_character_list2 = as(list_stringset_obj2, "CharacterList")
      as.list(as_character_list2)
      
      #Perform a pairwise sequence alignment for the two strings
      pairwise_align <- pairwiseAlignment(pattern = c(as_character_list2), subject = as_character_list)
      
      #returns the two aligned sequences aligned with pairwise sequence               alignment and the pairwise score
      return(pairwise_align)
}

# The function seq_pairwise performs pairwise sequence alignment for human and chimps when taking in the protein string set and the length of the sequences.
human_chimps_pairwise <- seq_pairwise(human_chimps_fasta, human_chimps_length)

# Here, I visualize the output of pairwise sequence alignment, which is the two aligned sequences, using the BrowseSeqs function. BrowseSeqs function takes in the pairwise sequence alignment result from above and the patterns argument indicates that the aligned sequences is a protein string set.  
human_chimps_plot <- BrowseSeqs(human_chimps_pairwise, patterns = AAStringSet)

# For my scientific hypothesis, I need to determine the percent similarity between human-chimps p53, human-dogs p53, and human-elephants p53. 

# The function seqidentity is used to determine the percent identity between two sequences and takes in either sequence alignment obtained from read.fasta or an alignment character matrix. 'ali' is an alignment character matrix. 

#Here, I get the alignment character matrix of human_chimps_read and stored it in the variable called human_chimps_aligned_seq. With the alignment character matrix I just found, I can pass it on to seqidentity. The result of seqidentity is the percent identity of the two sequences and I stored the percent identity in the variable human_chimps_identity.
human_chimps_aligned_seq <- human_chimps_read[["ali"]]
human_chimps_identity <- seqidentity(human_chimps_aligned_seq)
human_chimps_identity


# Here, I perform pairwise sequence alignment for human and dogs p53 sequences. The way to perform pairwise sequence alignment here for human and dogs is exactly the same as human and chimps. In order to reduce redundant commenting, please look at the detailed explanation on how to perform pairwise sequence alignment from above. 
human_dogs_fasta <- readAAStringSet("human_dogs.fasta.txt")
human_dogs_read <- read.fasta("human_dogs.fasta.txt")
human_dogs_length <- 1:(length(human_dogs_read)-1)

human_dogs_pairwise <- seq_pairwise(human_dogs_fasta, human_dogs_length)
human_dogs_plot <- BrowseSeqs(human_dogs_pairwise, patterns = AAStringSet)

human_dogs_aligned_seq <- human_dogs_read[["ali"]]
human_dogs_identity <- seqidentity(human_dogs_aligned_seq)
human_dogs_identity


# Here, I perform pairwise sequence alignment for human and elephants p53 sequences. The way to perform pairwise sequence alignment here for human and elephants is exactly the same as human and chimps. In order to reduce redundant commenting, please look at the detailed explanation on how to perform pairwise sequence alignment from above.
human_elephants_fasta <- readAAStringSet("human_elephants.fasta.txt")
human_elephants_read <- read.fasta("human_elephants.fasta.txt")
human_elephants_length <- 1:(length(human_elephants_read)-1)

human_elephants_pairwise <- seq_pairwise(human_elephants_fasta, human_elephants_length)
human_elephants_plot <- BrowseSeqs(human_elephants_pairwise, patterns = AAStringSet)

human_elephants_aligned_seq <- human_elephants_read[["ali"]]
human_elephants_identity <- seqidentity(human_elephants_aligned_seq)
human_elephants_identity

# In order to answer my scientific question, I need to analyze the three percent identities of p53 of chimps, dogs, and elephants to human p53. I want to answer that would animals that are known to be less susceptible to cancer have similar p53 sequences as humans since p53 is an important tumor suppressor protein in humans. In order to compare the three percent identities side by side, the data analysis method I chose is a histogram. 

# Histogram is a graphical representation of data that summarize the distribution of a data set. This method is great to be used here because it allows me to compared the percent identities of p53 of chimps, dogs, and elephants to humans side by side. 

# Here, I am creating a vector containing human_chimps_identity, human_dogs_identity, and human_elephants_identity in order to display them on the same plot. I stored the vector in the variable called percent_identity_all.
percent_identity_all <- c(human_chimps_identity, human_dogs_identity, human_elephants_identity)

# Here, I construct a histogram using the function hist. The histogram takes in the vector I created from above. Histogram takes in multiple arguments such as the main title, the axis title, and the color. There are more arguments that can be passed to make the histogram look more sophisticated, but the arguments I passed here is sufficient to create a histogram visualizing the percent identities of p53 of chimps, dogs, and elephants to humans side by side. 
hist(percent_identity_all,
main="percent identity of p53 of chimpanzees, dogs, and elephants to human p53",
xlab="Identity", xlim=c(0,1), col="blue",freq=FALSE)



```

Analyzing the Results:

Based on my understanding, chimpanzee and elephants are known to be less susceptible to cancer while dogs get cancer at a similar rate as humans. Because p53 is an important tumor-suppressor protein in human that reduces cancer, my scientific hypothesis is that if an animal is known to be less susceptible to cancer, then it will contain similar p53 sequence as humans. In other words, based on my hypothesis, chimpanzee and elephants will contain similar p53 sequence as humans because they are less susceptible to cancer while dogs will contain less similar p53 sequence as humans. My result of multiple sequence alignment shows that chimpanzee has high p53 sequence similarity to humans, but dogs and elephants have low p53 sequence similarity to humans. My result of pairwise sequence alignment shows similar results as multiple sequence alignment. Specifically, the percent identity of p53 for human and chimpanzee is 99%, 80% for humans and dogs, and 80% for humans and elephants. chimpanzee's p53 has high percent similarity to human p53, which is understandable because chimpanzee are less susceptible to cancer. Dogs' p53 has lower percent similarity to human p53, which is also understandable because dogs are more susceptible to cancer. However, elephants' p53 have low percent similarity to human p53 even though elephants are known to be less susceptible to cancer. Elephant's data suggests that my hypothesis cannot be confirmed to be correct at this time. The sample size for my research is too small, so more samples has to be studied before I am able to conclude my hypothesis with confidence. 
